File: Algorithms/0993_Cousins_in_Binary_Tree/Python/Cousins_in_Binary_Tree_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isCousins(self, root, x, y):
        x_level, y_level, x_parent, y_parent = -1, -1, None, None
        queue = [[root, 0, None]]
        while queue:
            cur_root, cur_level, cur_parent = queue.pop(0)
            if cur_root.val == x:
                x_level = cur_level
                x_parent = cur_parent
            if cur_root.val == y:
                y_level = cur_level
                y_parent = cur_parent
            if x_level != -1 and y_level != -1:
                return x_level == y_level and x_parent != y_parent
            if cur_root.left: queue.append([cur_root.left, cur_level + 1, cur_root])
            if cur_root.right: queue.append([cur_root.right, cur_level + 1, cur_root])
        return False

File: Algorithms/0993_Cousins_in_Binary_Tree/Java/CousinsinBinaryTreeSolution1.java

// Space: O(n)
// Time: O(n)


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<List<TreeNode>> queue = new ArrayDeque<>();
        List<TreeNode> temp = new ArrayList<>();
        temp.add(root);
        temp.add(null);
        queue.add(temp);
        TreeNode xParent = null, yParent = null;
        int xLevel = -1, yLevel = -1;
        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                List<TreeNode> cur = queue.poll();
                TreeNode curNode = cur.get(0);
                TreeNode curParentNode = cur.get(1);
                if (curNode.val == x) {
                    xParent = curParentNode;
                    xLevel = level;
                }
                if (curNode.val == y) {
                    yParent = curParentNode;
                    yLevel = level;
                }
                if (xLevel != -1 && yLevel != -1) {
                    return xParent != yParent && xLevel == yLevel;
                }
                if (curNode.left != null) {
                    ArrayList<TreeNode> left = new ArrayList<>();
                    left.add(curNode.left);
                    left.add(curNode);
                    queue.add(left);
                }
                if (curNode.right != null) {
                    ArrayList<TreeNode> right = new ArrayList<>();
                    right.add(curNode.right);
                    right.add(curNode);
                    queue.add(right);
                }
            }
            level++;
        }
        return false;
    }
}




